### PR TITLE
Glimmer GRPO: drop the planner kwargs the loader can derive itself

### PR DESCRIPTION
**Pull request by danielhanchen**
`#355` left the Glimmer GRPO notebook passing four `device_map_planner_kwargs`. Three of them are things the loader can work out, and one of them was wrong.

```python
device_map_planner_kwargs = {
    "rows_per_chunk": 128,                      # matches the log-softmax cap below
    "retained_rows": BATCH_SIZE * max_seq_length,
    "softcapped": True, "temperature_scaled": True,
},
```

- `softcapped` was already redundant. The planner reads `final_logit_softcapping` off the config when the flag is left unset, and it already looks inside `text_config`, which is where Glimmer's `20.0` lives.
- `rows_per_chunk: 128` is the planner's own default.
- `temperature_scaled: True` was actively wrong. It is a GRPO hyperparameter, not a model fact, and the notebook sets `temperature = 1.0`, which is also TRL's default (checked at `v0.22.2` and at the pinned `v1.9.2`). The chunked log-softmax guards the transform with `if temperature != 1.0`, so it never runs. The notebook was reserving 103,448,576 bytes (98.66 MiB) for a temporary that is never allocated.

`retained_rows` stays. It defaults to 0, which models an inference forward under `no_grad`; a GRPO backward keeps every chunk alive until backward runs, so the loader cannot derive it.

Result:

```python
# Only what the loader cannot know. The planner reads the soft cap off
# text_config.final_logit_softcapping, and rows_per_chunk already defaults
# to the 128 used below. retained_rows has to be passed: it defaults to 0,
# which models inference under no_grad, and a GRPO backward keeps every
# chunk alive until backward.
device_map_planner_kwargs = {"retained_rows": BATCH_SIZE * max_seq_length},
```

## No merge-ordering constraint

An earlier revision of this said the logit scale is read off the config too, as of `unslothai/unsloth-zoo#1067`. That was wrong twice over, and the comment no longer claims it:

- #1067 is still open, so no shipped planner auto-detects a logit scale. `unsloth_zoo` main and PyPI `2026.8.12` are the same content today, and both have `logit_scaled: bool = False`.
- Even with #1067 applied, Glimmer would not be covered. Its scale field is `text_config.output_multiplier` (`0.196...`), and #1067 matches only `logit_scale`, `lm_head_multiplier` and `logits_scaling`.

Neither point affects this change, because the notebook never passed `logit_scaled` in the first place, before or after. Soft-cap detection is the only auto-detection this leans on, and that has been in zoo main all along, `text_config` included.

Checked by planning `unsloth/Muse-Glimmer-30B-unsloth-bnb-4bit` on the meta device with fabricated budgets, against both the current zoo main planner and the same file with #1067 applied:

| shape | device map, before vs after | headroom delta |
|---|---|---|
| 2 x 16 GiB (the Kaggle T4 pair) | identical | -103,448,576 B |
| 2 x 24 GiB / 2 x 40 GiB / 2 x 80 GiB | identical | -103,448,576 B |
| 4 x 16 GiB, and 16+24 GiB asymmetric | one decoder layer relocated, see below | -103,448,576 B |
| 1 GPU, and CPU only | planner declines on both, unchanged | n/a |

and on every shape the post-change plan is byte-identical between the pre-#1067 and post-#1067 planner. So this can merge in any order relative to #1067.

For scale: if soft-cap detection did not cover `text_config`, headroom would fall by 1.2033 GiB rather than 98.66 MiB. It does cover it, on both planners.

### The two shapes whose map moves are not a regression

Freeing 103,448,576 B of reserve on the head's card changes the slack the balanced packing walk starts from, so on the two shapes with more than two cards to balance it settles on a different, equally valid assignment. In both cases exactly one decoder layer of 242,012,160 B moves, total weights are unchanged, and every card still fits:

| | 4 x 16 GiB | 16+24 GiB |
|---|---|---|
| head card | gpu3, unchanged | gpu1, unchanged |
| total weight bytes | 21,807,663,808 before and after | 21,807,663,808 before and after |
| relocated unit | `layers.26`, gpu1 -> gpu2 | `layers.12`, gpu0 -> gpu1 |
| tightest card after | gpu2, 96,906,416 B spare | gpu1, 109,262,279 B spare |

Neither shape is a configuration this notebook sizes for; the sizing cell targets one card or two. They are here to show the change does not over-commit a card when more cards are present.

### Why the notebook runs transformers 5.15.0

Worth stating plainly, since the version numbers do not line up on their own. `unsloth`'s `pyproject.toml` declares `transformers>=4.51.3,...,<=5.5.0`, and the notebook installs `transformers==5.15.0`, which is outside that range. That is deliberate and unavoidable: `muse_glimmer` does not exist in transformers `5.5.0` at all, so Glimmer cannot be loaded anywhere inside the declared pin. The notebook installs transformers last and asserts the version, so `pip` reports the conflict and completes, and `unsloth` does not enforce the ceiling at import (its import-time check validates transformers' own dependency floors and only warns).

This predates the branch and applies to the whole Glimmer notebook family, not just GRPO. It does not touch the ordering verdict above, which rests only on zoo's soft-cap detection and on the notebook never passing `logit_scaled`.

## Notes

Edited `original_template/Muse_Glimmer_(30B)-GRPO.ipynb` and re-ran the generators with the repo's pinned `marimo==0.23.8` and `ruff==0.15.16`, then reverted the unrelated churn a full regeneration produces (`scripts/model_created_at.csv` re-fetches its timestamps, which reorders the README). A control regeneration with no source edit leaves all eight Glimmer files untouched, and produces that same unrelated churn on the base branch, so it is pre-existing.

Unrelated finding, recorded so it is not lost: the GRPO loss side derives `logit_scale_multiply` from `logit_scale` / `logits_scaling` only, so Glimmer's `output_multiplier` is never applied to the chunked logits. That is an `unsloth` issue rather than a notebook one, and it predates this branch.

Opened against `#355`'s branch since it introduced the block being edited. Retarget to `main` once that merges.

### CODE REVIEW

**Comment by danielhanchen** [comment]
Auto-closed when its base branch was deleted on merging #355, and it cannot be reopened after the rebase onto main. Continued in #357 with the same content, rebased with no conflicts.
